src/terrain/terrain_base.cpp

```diff
@@ -0,0 +1,223 @@
+            auto mat_color = materials_inverse.at(color);
+            tiles_.push_back(Tile({x, y, z}, mat_color.first, mat_color.second));
+        } else { // the color is unknown
+            unknown_materials.insert(color);
+            auto mat_color = materials_inverse.at(0); // else set to air.
+            tiles_.push_back(Tile({x, y, z}, mat_color.first, mat_color.second));
+        }
+    }
+
+    for (uint32_t color : unknown_materials) {
+        // is there any way to log hexadecimal
+        LOG_WARNING(logging::terrain_logger, "Cannot find color: {:x}", color);
+    }
+}
+
+int
+TerrainBase::get_first_not(
+    const std::set<std::pair<int, int>> materials, int x, int y, int guess
+) const {
+    if (guess < 1) {
+        guess = 1;
+    }
+    if ((uint32_t)guess >= Z_MAX) {
+        guess = Z_MAX - 1;
+    }
+    if (materials.find(std::make_pair(
+            get_tile(x, y, guess - 1)->get_material_id(),
+            get_tile(x, y, guess - 1)->get_color_id()
+        ))
+        != materials.end()) {
+        if (materials.find(std::make_pair(
+                get_tile(x, y, guess)->get_material_id(),
+                get_tile(x, y, guess)->get_color_id()
+            ))
+            == materials.end()) {
+            return guess;
+        } else {
+            // go up
+            for (uint16_t z = guess + 1; z < Z_MAX; z++) {
+                if (materials.find(std::make_pair(
+                        get_tile(x, y, z)->get_material_id(),
+                        get_tile(x, y, z)->get_color_id()
+                    ))
+                    == materials.end()) {
+                    return z;
+                }
+            }
+            return Z_MAX; // -1?
+        }
+    } else {
+        // go down
+        for (uint16_t z = guess - 2; z > 0; z--) {
+            if (materials.find(std::make_pair(
+                    get_tile(x, y, z)->get_material_id(),
+                    get_tile(x, y, z)->get_color_id()
+                ))
+                != materials.end()) {
+                return z + 1;
+            }
+        }
+        return 0;
+    }
+}
+
+void
+TerrainBase::add_to_top(
+    Json::Value top_data, const std::map<uint8_t, const Material>* materials
+) {
+    std::set<std::pair<int, int>> material_type;
+
+    for (Json::Value::ArrayIndex i = 0; i < top_data["above_colors"].size(); i++) {
+        int E = top_data["above_colors"][i]["E"].asInt();
+        if (top_data["above_colors"][i]["C"].isInt()) {
+            int C = top_data["above_colors"][i]["C"].asInt();
+            material_type.insert(std::make_pair(E, C));
+        } else if (top_data["above_colors"][i]["C"].asBool()) {
+            for (unsigned int C = 0; C < (*materials).at(E).color.size(); C++) {
+                material_type.insert(std::make_pair(E, C));
+            }
+        }
+    }
+
+    uint16_t guess = 0;
+    // for loop
+    for (uint16_t x = 0; x < X_MAX; x++)
+        for (uint16_t y = 0; y < Y_MAX; y++) {
```

Comment:
Induction vars are size_t

Reply:
clearly they are `uint16_t`.

Reply:
There is no advantage to using a `uint16_t` for an induction variable. The register it uses is still 64 bits wide.